actors/miner/src/lib.rs

```diff
@@ -977,188 +988,213 @@ impl Actor {
+        rt.validate_immediate_caller_is(
+            info.control_addresses.iter().chain(&[info.owner, info.worker]),
+        )?;
+        if params.sector_proofs.is_empty() == params.aggregate_proof.is_empty() {
+            return Err(actor_error!(
+                illegal_argument,
+                "exactly one of sector proofs or aggregate proof must be non-empty"
+            ));
+        }
 
-                deadline.partitions =
-                    partitions.flush().with_context_code(ExitCode::USR_ILLEGAL_STATE, || {
-                        format!("failed to save partitions for deadline {}", dl_idx)
-                    })?;
+        // Load sector infos for validation, failing if any don't exist.
+        let mut sectors = Sectors::load(&store, &state.sectors)
+            .context_code(ExitCode::USR_ILLEGAL_STATE, "failed to load sectors array")?;
+        let mut sector_infos = Vec::with_capacity(params.sector_updates.len());
+        let mut updates = Vec::with_capacity(params.sector_updates.len());
+        for (i, update) in params.sector_updates.iter().enumerate() {
+            sector_infos.push(sectors.must_get(update.sector)?);
+            let sector_type = sector_infos.last().unwrap().seal_proof;
+            let computed_commd =
+                unsealed_cid_from_pieces(rt, &update.pieces, sector_type)?.get_cid(sector_type)?;
+
+            updates.push(ReplicaUpdateInner {
+                sector_number: update.sector,
+                deadline: update.deadline,
+                partition: update.partition,
+                new_sealed_cid: update.new_sealed_cid,
+                new_unsealed_cid: Some(computed_commd),
+                deals: vec![],
+                update_proof_type: params.update_proofs_type,
+                // Replica proof may be empty if an aggregate is being proven.
+                // Validation needs to accept this empty proof.
+                replica_proof: params.sector_proofs.get(i).unwrap_or(&RawBytes::default()).clone(),
+            });
+        }
 
-                deadlines
-                    .update_deadline(rt.policy(), rt.store(), dl_idx, &deadline)
-                    .with_context_code(ExitCode::USR_ILLEGAL_STATE, || {
-                        format!("failed to save deadline {}", dl_idx)
-                    })?;
-            } // End loop over deadlines
+        // Validate inputs.
+        let require_deals = false; // No deals can be specified in new replica update.
+        let (batch_return, update_sector_infos) = validate_replica_updates(
+            &updates,
+            &sector_infos,
+            &state,
+            info.sector_size,
+            rt.policy(),
+            rt.curr_epoch(),
+            store,
+            require_deals,
+            params.require_activation_success,
+        )?;
+        let valid_unproven_usis = batch_return.successes(&update_sector_infos);
+        let valid_original_updates = batch_return.successes(&params.sector_updates);
```

Comment:
Renamed to use "manifest" for those ones. Yes there might be a path to unifying. I think it will be a little easier after delete the old pru2, since the current ReplicaUpdateInner exists to abstract between the entry points.